File: DSA/Yashvardhan-Guliyani/Day17_q1.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    
    TreeNode(int val) {
        this.val = val;
    }
}

public class BSTLowestCommonAncestor {
    
    public static TreeNode lowestCommonAncestor(TreeNode root, int p, int q) {
        if (root == null) {
            return null;
        }
    
        if (p > root.val && q > root.val) {
            return lowestCommonAncestor(root.right, p, q);
        }
        if (p < root.val && q < root.val) {
            return lowestCommonAncestor(root.left, p, q);
        }
        return root;
    }
    
    public static TreeNode insertIntoBST(TreeNode root, int val) {
        if (root == null) {
            return new TreeNode(val);
        }
        
        if (val < root.val) {
            root.left = insertIntoBST(root.left, val);
        } else if (val > root.val) {
            root.right = insertIntoBST(root.right, val);
        }
        
        return root;
    }
    
    public static void main(String[] args) {
        TreeNode root = null;
        int[] values = {20, 10, 30, 5, 15, 25, 35};
        
        for (int val : values) {
            root = insertIntoBST(root, val);
        }
        int node1 = 5;
        int node2 = 30;
        
        TreeNode lca = lowestCommonAncestor(root, node1, node2);
        
        if (lca != null) {
            System.out.println("The Lowest Common Ancestor of " + node1 + " and " + node2 + " is: " + lca.val);
        } else {
            System.out.println("LCA not found or one of the nodes doesn't exist in the tree.");
        }
    }
}

File: DSA/Aarohi-Sharma/Day17_q1.cpp
#include <iostream>
using namespace std;
class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root) return NULL;
        if (root->val < p->val && root->val < q->val) {
            return lowestCommonAncestor(root->right, p, q);
        }
        if (root->val > p->val && root->val > q->val) {
            return lowestCommonAncestor(root->left, p, q);
        }
        return root;
    }
};
int main() {
    TreeNode* root = new TreeNode(20);
    root->left = new TreeNode(10);
    root->right = new TreeNode(30);
    root->left->left = new TreeNode(5);
    root->left->right = new TreeNode(15);
    root->right->left = new TreeNode(25);
    root->right->right = new TreeNode(35);
    Solution solution;
    TreeNode* p = root->left->left;  
    TreeNode* q = root->right; 
    TreeNode* lca = solution.lowestCommonAncestor(root, p, q);
    if (lca) {
        cout << "The LCA of nodes " << p->val << " and " << q->val << " is " << lca->val << endl;
    } else {
        cout << "LCA not found." << endl;
    }
    return 0;
}
